### DISCUSSION TITLE
How to skip/ignore down hosts

### DISCUSSION THREAD

**Discussion by barmar** [deployphp/deployer#2883] [Help needed]
We have a dozen production hosts in our deploy.php, and sometimes some of them will be down. If we try to do a deployment during this time, it will fail and abort everything. Is there a way to have Deployer ignore these servers and complete the deployment to the working servers? And then when they've recovered, get them caught up?

I guess we could do it by listing the machines to deploy to by hand on the command line, rather than the stage that includes all of them. But is there something more automatic?

We're currently using version 6.8.0, so if the solution requires upgrading to 7.x let me know.

**Comment by antonmedv** [comment]
Currently Deployer doesn’t have such ability. I think it’s better suited to be configured outside of Deployer. 

In v7 you can do something like this:

    dep deploy all,host!=beta

**Reply by barmar** [reply]
> In v7 you can do something like this:
> 
> ```
> dep deploy all,host!=beta
> ```

I'm not sure how that helps. It's not a predefined group of hosts, it's just hosts that happen to be down at the moment, e.g. for maintenance or unexpected failure.

**Reply by antonmedv** [reply]
Figure out down host before, via some script. Call deployer with selector for only online hosts.

**Reply by TimWuyts** [reply]
I'm looking for the same solution. I had already created an additional task to check if the server is available. When not the GracefulShutdownException is triggered, but this is stopping the deployment process for all hosts.

```
namespace Deployer;

use Deployer\Exception\GracefulShutdownException;

task('deploy:check_host', function () {
    $hostname = get('hostname');
    $command = "curl -Is " . escapeshellarg('https://' . $hostname) . " | head -1";
    $output = shell_exec($command);

    // cancel deploy when curl request returns a 503 server error
    if (strpos($output, '503') > 0) {
        throw new GracefulShutdownException("Host is not available. Skipping.");
    }
});
```

**Reply by antonmedv** [reply]
Maybe let's add `IgnoreHostException` to ignore current host?

**Reply by TimWuyts** [reply]
That would be a great addition in my opinion!

**Reply by antonmedv** [reply]
Please create an issue so we will remember to do this)
